appcues/src/main/java/com/appcues/data/model/Experience.kt

```diff
@@ -48,4 +54,37 @@ internal data class Experience(
             }
         }
     }
+
+    // Gets any actions defined on the step group container for the "navigate" trigger. These are the
+    // actions that should be executed before presenting this group's container.
+    fun getNavigationActions(stepContainerIndex: Int): List<ExperienceAction> {
+        val stepGroup = stepContainers.getOrNull(stepContainerIndex) ?: return arrayListOf()
+        return stepGroup.actions[stepGroup.id]?.filter { it.on == NAVIGATE }?.map { it.experienceAction } ?: emptyList()
+    }
+
+    fun areStepsFromDifferentGroup(stepIndexOne: Int, stepIndexTwo: Int): Boolean {
+        return groupLookup[stepIndexOne] != groupLookup[stepIndexTwo]
+    }
+
+    fun getPresentingTrait(flatStepIndex: Int): PresentingTrait {
```

Comment:
the old implementation had a fundamental problem, it is not fully testable since we know internally both lists are coupled together, but whenever accessing a list (stepContainers) we need to test what would happen if we get IndexOutOfBounds, or replace it by getOrNull to ensure all cases are covered.

so changing this code seems logical as we should avoid the possibility of untestable state in general

```kotlin
internal fun BeginningStep.presentingTrait(): PresentingTrait? {
    return experience.groupLookup[flatStepIndex]?.let { experience.stepContainers[it].presentingTrait }
}
```

Reply:
previously, did we have error handling for the null return value for `PresentingTrait?` and throwing an AppcuesTraitException somewhere? Want to make sure we don't change the error handling behavior, as this is now possibly throwing a different exception type.

Reply:
reusing bit of code since the same exception is used inside Experience

Reply:
actually took another approach here were we dont need to change AppcuesTraitException to "open"